### PR TITLE
Implement JWT Refresh Token

### LINKED ISSUE #78: [AUTH] Keep sessions alive
Currently, we do not have a refresh token so we have no way to keep a session alive for a user. We should add that refresh token so a user doesn't have to be constantly logging in.

### LINKED ISSUE #86: [FRONTEND] Style the login page
The login page needs to look nice. Include the logo above the login field.

### LINKED ISSUE #87: [AUTH] admin password file
Currently, auth is checked against a hard coded user and password. 

Unless someone can give a good reason, I don't think we need a username for the app. There should be one admin for Rainstorm. Other users will be at the service level and maybe we can allow the admin to create users later. But those users dont need to manage the services themselves.

Since theres only one admin, the user name cam be ommitted.

Since we're not using a DB, the password hash will be stored somewhere in the filesystem. If it is stored on the SD card, it will be reset when the user upgrades, unless premium. This could be an intentional headache to promote premium purchases. Otherwise, store the password hash file on the HDD.

### LINKED ISSUE #101: [AUTH] allow change password from settings

### PR DESCRIPTION
**Pull request by asharron**
# What was done 
- Updated our JWT library to be flask_jwt_extended since it came with refresh tokens out of the box
- Updated the login route to return an access_token and a refresh_token 
- Updated login logic to read from the app config for username and password
- Added a refresh api call method to the AuthService class
- Updated logic for logout button to delete tokens from storage instead of just sending you to the login screen
- Updated Login component to automatically try and use a refresh token to re-authenticate if one is present 

# How To Test
1. Checkout the branch on your pi 
2. Within your python virutal env, run pip3 install -r requriements.txt again to download flask_jwt_extended
2. Edit line 24 of __init__.py for the flask server to be 5 seconds instead of 900 (so you don't have to wait 15 min to verify the refresh logic)
```
JWT_ACCESS_TOKEN_EXPIRES=timedelta(seconds=900),  # 15 min
```
2. Start FED and BED dev servers 
3. Logout on the frontend (in case you already have an access_token stored)
4.  Login using the drop user name and password 
5. Wait 5 seconds 
6. Hit refresh 
7. You should still be logged in
8. Verify the server logs show you were unauthenticated but it automatically refreshed your access_token

image

### CODE REVIEW

File: raincloud/__init__.py [hunk 1]
```diff
@@ -5,23 +5,28 @@
...
     # create and configure the app
     app = Flask(__name__, instance_relative_config=True)
-    jwt = JWT(app, authenticate, identity)
+    jwt = JWTManager(app)
     app.config.from_mapping(
         SECRET_KEY='dev',
         JWT_SECRET_KEY='secret',
-        JWT_EXPIRATION_DELTA=timedelta(seconds=86400),
+        JWT_ACCESS_TOKEN_EXPIRES=timedelta(seconds=900),  # 15 min
+        JWT_REFRESH_TOKEN_EXPIRES=timedelta(seconds=2592000),  # 30 days
```
Comment (WillHHippo) [inline raincloud/__init__.py]: This should probably be like a day at most. Keeping someone logged in for 30 days doesn't seem great.

**Review by WillHHippo — CHANGES_REQUESTED** [review]
Small change requested, otherwise looks good.

**Comment by nullcount** [comment]
Closes #78 #87 #86 #101

**Comment by count-null** [comment]
Bug found when testing. 

Steps to reproduce:
1. Start up app from scratch
2. Launch web browser
3. Notice that content does not load

Desired behaviour: 
1. App should notice that user is not logged in and redirect to the login page

Thoughts: 
Maybe the refresh token logic is missing the initial case where the user has never logged in at all. i.e. there is no token to begin with.

**Comment by nullcount** [comment]
I added some commits to enable changing password. Also styled the login page a bit. Added nice alerts to indicate errors, success in the login and change password process.

File: raincloud/__init__.py (hunk 1 above)
Comment (asharron) [reply raincloud/__init__.py]: Updated :+1:

**Pull request merged**